src/main.rs

```diff
@@ -65,9 +64,18 @@ fn main() -> Result<()> {
     if let Some(filename) = args.filename {
         let mut has_newline = true;
 
-        if filename.exists() {
-            let content = fs::read(&filename)
-                .with_context(|| anyhow!("Failed to open file: {:?}", filename))?;
+        let handle = match File::open(&filename) {
+            Ok(file) => Ok(Some(file)),
+            Err(error) => match error.kind() {
+                ErrorKind::NotFound => Ok(None),
+                _ => Err(error)
+            },
+        }.with_context(|| anyhow!("Failed to open file: {:?}", filename))?;
```

Comment:
You could probably use something like this to avoid a TOCTOU, but still keep the optimized `fs::read`:

```rust
        match fs::read(&filename) {
            Ok(content) => {
                // initialize `set` from `content`
                // ...
            },
            Err(error) => match error.kind() {
                ErrorKind::NotFound => (),
                _ => return Err(error.context(anyhow!("Failed to read file: {:?}", filename))),
            },
        }
```

Reply:
Thanks for the suggestions. Applied.